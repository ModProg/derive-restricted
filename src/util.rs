//! Utility functions.

use core::iter::FromIterator;

use proc_macro2::Span;
use syn::{punctuated::Punctuated, Ident, Path, PathArguments, PathSegment, Token};

use crate::Item;

/// Create [`PathSegment`] from [`str`].
pub fn path_segment(ident: &str) -> PathSegment {
    PathSegment {
        ident: Ident::new(ident, Span::call_site()),
        arguments: PathArguments::None,
    }
}

/// Create [`Path`] from `[&str]`.
pub fn path(segments: &[&str]) -> Path {
    Path {
        leading_colon: Some(<Token![::]>::default()),
        segments: Punctuated::from_iter(segments.iter().map(|segment| path_segment(segment))),
    }
}

/// Return if given enum has any empty or unit variants. If not an enum, will always return `false`.
pub fn unit_found(data: &Item) -> bool {
    if let Item::Enum { variants, .. } = data {
        variants.iter().any(|variant| variant.fields.is_empty())
    } else {
        false
    }
}
