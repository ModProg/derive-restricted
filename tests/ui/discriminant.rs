use std::marker::PhantomData;

use derive_where::derive_where;

#[derive_where(PartialEq, PartialOrd)]
#[repr(A)]
enum UnknownRepr<T> {
	A(PhantomData<T>),
}

#[derive_where(PartialEq, PartialOrd)]
enum MissingIntegerRepr<T> {
	A(PhantomData<T>) = 0,
}

#[derive_where(PartialEq, PartialOrd)]
#[repr(C)]
enum MissingIntegerReprC<T> {
	A(PhantomData<T>) = 0,
}

fn main() {}
