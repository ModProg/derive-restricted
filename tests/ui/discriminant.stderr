error[E0552]: unrecognized representation hint
 --> tests/ui/discriminant.rs:6:8
  |
6 | #[repr(A)]
  |        ^
  |
  = help: valid reprs are `Rust` (default), `C`, `align`, `packed`, `transparent`, `simd`, `i8`, `u8`, `i16`, `u16`, `i32`, `u32`, `i64`, `u64`, `i128`, `u128`, `isize`, `usize`

error[E0732]: `#[repr(inttype)]` must be specified
  --> tests/ui/discriminant.rs:12:1
   |
12 | enum MissingIntegerRepr<T> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0732]: `#[repr(inttype)]` must be specified
  --> tests/ui/discriminant.rs:18:1
   |
18 | enum MissingIntegerReprC<T> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
